### PR TITLE
[Sol->Yul] Implement keccak & getter functions

### PR DESCRIPTION
**Pull request by Marenz**
Part of #8343.

### CODE REVIEW

File: libsolidity/codegen/ir/IRGenerationContext.cpp [hunk 1]
```diff
@@ -65,6 +65,13 @@ string IRGenerationContext::functionName(FunctionDefinition const& _function)
 	return "fun_" + _function.name() + "_" + to_string(_function.id());
 }
 
+string IRGenerationContext::functionName(VariableDeclaration const& _varDecl)
+{
+	// @TODO previously, we had to distinguish creation context and runtime context,
+	// but since we do not work with jump positions anymore, this should not be a problem, right?
+	return "fun_" + _varDecl.name() + "_" + to_string(_varDecl.id());
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerationContext.cpp]: maybe better to use `getter_fun_`?

Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: Also I think that todo can be deleted.

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 2]
```diff
@@ -149,6 +149,25 @@ string IRGenerator::generateFunction(FunctionDefinition const& _function)
 	});
 }
 
+string IRGenerator::generateGetter(VariableDeclaration const& _varDecl)
+{
+	string functionName = m_context.functionName(_varDecl);
+
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerator.cpp:155]: Please assert that it is a storage variable that is not a constant.

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 3]
```diff
@@ -234,7 +253,14 @@ string IRGenerator::dispatchRoutine(ContractDefinition const& _contract)
 		templ["abiDecode"] = abiFunctions.tupleDecoder(type->parameterTypes());
 		templ["params"] = suffixedVariableNameList("param_", 0, paramVars);
 		templ["retParams"] = suffixedVariableNameList("ret_", retVars, 0);
-		templ["function"] = generateFunction(dynamic_cast<FunctionDefinition const&>(type->declaration()));
+
+		if (FunctionDefinition const* funDef = dynamic_cast<FunctionDefinition const*>(&type->declaration()))
+			templ["function"] = generateFunction(*funDef);
+		else if(VariableDeclaration const* varDecl = dynamic_cast<VariableDeclaration const*>(&type->declaration()))
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerator.cpp]:
```suggestion
		else if (VariableDeclaration const* varDecl = dynamic_cast<VariableDeclaration const*>(&type->declaration()))
```

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerator.cpp:155]: Please unimplementedAssert that it is a value type.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 4]
```diff
@@ -607,6 +607,26 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 
 		break;
 	}
+	case FunctionType::Kind::KECCAK256:
+	{
+		solAssert(arguments.size() == 1, "");
+
+		ArrayType const* arrayType = TypeProvider::bytesMemory();
+		string const& array = m_context.newYulVariable();
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp:615]: Is there a benefit of using `const&` here?

File: test/libsolidity/semanticTests/viaYul/keccak.sol [hunk 5]
```diff
@@ -0,0 +1,14 @@
+contract C {
+    function keccak1() public pure returns (bytes32) {
+		return keccak256("123");
+    }
+    function keccak2() public pure returns (bytes32) {
+		bytes memory a = "456";
```
Comment (chriseth) [inline test/libsolidity/semanticTests/viaYul/keccak.sol]: Can you also use `123` here?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 4 above)
Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:615]: We never want nor should modify it after creation, the `const` assures this. There might also be optimization advantages.

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:615]: ok, but what about the `&`?

Comment (Marenz) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:615]: I suppose it's not doing a copy of the data either way.. Do you see a disadvantage to it?

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `100eded91f507f5a058b6ec524029b33f22bd363`:
```
Coding style error:
 libsolidity/codegen/ir/IRGenerator.cpp:175: });

```
Please check that your changes are working as intended.

**Comment by Marenz** [comment]
Updated

File: libsolidity/codegen/ir/IRGenerator.cpp [hunk 6]
```diff
@@ -149,6 +149,32 @@ string IRGenerator::generateFunction(FunctionDefinition const& _function)
...
 }
 
+string IRGenerator::generateGetter(VariableDeclaration const& _varDecl)
+{
+	string functionName = m_context.functionName(_varDecl);
+
+	Type const* type = _varDecl.annotation().type;
+
+	solAssert(!_varDecl.isConstant(), "");
+	solAssert(!type->dataStoredIn(DataLocation::Storage), "");
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerator.cpp]: I actually meant `_varDecl.isStateVariable()`

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp:615]: The disadvantage is that you have to think about whether `newYulVariable` might return via reference and the lifetime of that reference in that case.

File: libsolidity/codegen/ir/IRGenerator.cpp (hunk 6 above)
Comment (Marenz) [reply libsolidity/codegen/ir/IRGenerator.cpp]: ah. Fixed.

**Review by chriseth — APPROVED** [review]

**Pull request merged**